src/docker/restorableappengine.cc

```diff
@@ -0,0 +1,462 @@
+                                   [&dir](const AppEngine::App& app) { return dir == app.name; });
+
+    if (foundAppIt == app_shortlist.end()) {
+      // remove App dir tree since it's not found in the shortlist
+      boost::filesystem::remove_all(entry.path());
+      continue;
+    }
+
+    const auto& app{*foundAppIt};
+    const Uri uri{Uri::parseUri(app.uri)};
+
+    // iterate over `app` subdirectories/versions and remove those that doesn't match the specified version
+    const auto app_dir{apps_root_ / uri.app};
+
+    for (const auto& entry : boost::make_iterator_range(boost::filesystem::directory_iterator(app_dir), {})) {
+      if (!boost::filesystem::is_directory(entry)) {
+        LOG_WARNING << "Found file while expected an App version directory: " << entry.path().filename().native();
+        continue;
+      }
+
+      const std::string app_version_dir = entry.path().filename().native();
+      if (app_version_dir != uri.digest.hash()) {
+        boost::filesystem::remove_all(entry.path());
+        continue;
+      }
+
+      // add blobs of the shortlisted apps to the blob shortlist
+      ComposeInfo compose{(entry.path() / ComposeFile).string()};
+      for (const auto& service : compose.getServices()) {
+        const auto image = compose.getImage(service);
+        const Uri image_uri{Uri::parseUri(image)};
+        const auto image_root{app_dir / app_version_dir / "images" / image_uri.registryHostname / image_uri.repo /
+                              image_uri.digest.hash()};
+
+        const auto image_manifest_desc{Utils::parseJSONFile(image_root / "index.json")};
+        HashedDigest image_digest{image_manifest_desc["manifests"][0]["digest"].asString()};
+        blob_shortlist.emplace(image_digest.hash());
+
+        const auto image_manifest{Utils::parseJSONFile(blobs_root_ / "sha256" / image_digest.hash())};
+        blob_shortlist.emplace(HashedDigest(image_manifest["config"]["digest"].asString()).hash());
+
+        const auto image_layers{image_manifest["layers"]};
+        for (Json::ValueConstIterator ii = image_layers.begin(); ii != image_layers.end(); ++ii) {
+          if ((*ii).isObject() && (*ii).isMember("digest")) {
+            const auto layer_digest{HashedDigest{(*ii)["digest"].asString()}};
+            blob_shortlist.emplace(layer_digest.hash());
+          } else {
+            LOG_ERROR << "Invalid image manifest: " << ii.key().asString() << " -> " << *ii;
+          }
+        }
+      }
+    }
+  }
+
+  // prune blobs
+  if (!boost::filesystem::exists(blobs_root_ / "sha256")) {
+    return;
+  }
+
+  for (const auto& entry :
+       boost::make_iterator_range(boost::filesystem::directory_iterator(blobs_root_ / "sha256"), {})) {
+    if (boost::filesystem::is_directory(entry)) {
+      continue;
+    }
+
+    const std::string blob_sha = entry.path().filename().native();
+    if (blob_shortlist.end() == blob_shortlist.find(blob_sha)) {
+      LOG_DEBUG << "Removing blob: " << entry.path();
```

Comment:
After running this on my device, I'd say we do this LOG_INFO. It's pretty useful to know about.

Reply:
updated